src/Microsoft.Repl.Tests/ShellTests.cs

```diff
@@ -0,0 +1,191 @@
+            Assert.Equal(inputBufferTextAfterKeyPress, shell.ShellState.InputManager.GetCurrentBuffer());
+        }
+
+        [Fact]
+        public void RunAsync_WithEscapeKeyPress_UpdatesInputBufferWithEmptyString()
+        {
+            Shell shell = CreateShell(ConsoleKey.Escape,
+                0,
+                null,
+                null,
+                out CancellationTokenSource cancellationTokenSource);
+
+            string inputBufferTextBeforeKeyPress = "get";
+            string inputBufferTextAfterKeyPress = string.Empty;
+
+            IShellState shellState = shell.ShellState;
+            shellState.InputManager.SetInput(shellState, inputBufferTextBeforeKeyPress);
+
+            // Verify the input buffer contents before Escape key press event
+            Assert.Equal(inputBufferTextBeforeKeyPress, shell.ShellState.InputManager.GetCurrentBuffer());
+
+            shell.RunAsync(cancellationTokenSource.Token);
+
+            // Verify the input buffer contents after Escape key press event
+            Assert.Equal(inputBufferTextAfterKeyPress, shell.ShellState.InputManager.GetCurrentBuffer());
+        }
+
+        [Fact]
+        public void RunAsync_WithInsertKeyPress_FlipsIsOverwriteModeInInputManager()
+        {
+            Shell shell = CreateShell(ConsoleKey.Insert,
+                0,
+                null,
+                null,
+                out CancellationTokenSource cancellationTokenSource);
+
+            // Verify IsOverwriteMode flag in input manager is set to true before Insert key press event
+            Assert.False(shell.ShellState.InputManager.IsOverwriteMode);
+
+            shell.RunAsync(cancellationTokenSource.Token);
+
+            // Verify IsOverwriteMode flag in input manager is set to false after Insert key press event
+            Assert.True(shell.ShellState.InputManager.IsOverwriteMode);
+        }
+
+        [Fact]
+        public void RunAsync_WithUnhandledKeyPress_DoesNothing()
+        {
+            Shell shell = CreateShell(ConsoleKey.F1,
+                0,
+                null,
+                null,
+                out CancellationTokenSource cancellationTokenSource);
+
+            string inputBufferText = "get";
+
+            IShellState shellState = shell.ShellState;
+            shellState.InputManager.SetInput(shellState, inputBufferText);
+
+            // Verify the input buffer contents before F1 key press event
+            Assert.Equal(inputBufferText, shell.ShellState.InputManager.GetCurrentBuffer());
+
+            shell.RunAsync(cancellationTokenSource.Token);
+
+            // Verify the input buffer contents after F1 key press event
+            Assert.Equal(inputBufferText, shell.ShellState.InputManager.GetCurrentBuffer());
+        }
+
+        private Shell CreateShell(ConsoleKey consoleKey, int caretPosition, string previousCommand, string nextCommand, out CancellationTokenSource cancellationTokenSource)
```

Comment:
What do you think about splitting this function out a bit to make it more clear what's mocked and what's "real"? Right now the httpState, defaultCommandDispatcher, cancellationTokenSource/cancellationToken, shellState and the Shell itself are all "real". The IConsoleManager and ICommandHistory are mocked. I think it might be easier when we come back to this code to add more tests in a few months or whatever if we can at a glance see what's being mocked. I'm not positive its necessary though. Thoughts?

Reply:
I'll be sending out a pr with tests for the remaining (handled) key press events - LeftArrow, RightArrow, Home, End, Tab and Enter. From my investigation, they are not straightforward and I'm guessing this method needs to be revisited and tweaked to handle those scenarios. I will leave this as is now and will tweak it/refactor as required in my next pr.

Let me know if anyone else has any other suggestions.